src/main/java/htsjdk/samtools/util/AbstractBlockCompressedOutputStream.java

```diff
@@ -0,0 +1,262 @@
+ * bytes reaches a threshold.
+ *
+ * The advantage of BGZF over conventional gzip is that BGZF allows for seeking without having to scan through
+ * the entire file up to the position being sought.
+ *
+ * Note that the flush() method should not be called by client
+ * unless you know what you're doing, because it forces a gzip block to be written even if the
+ * number of buffered bytes has not reached threshold.  close(), on the other hand, must be called
+ * when done writing in order to force the last gzip block to be written.
+ *
+ * c.f. http://samtools.sourceforge.net/SAM1.pdf for details of BGZF file format.
+ */
+public abstract class AbstractBlockCompressedOutputStream extends OutputStream implements LocationAware {
+
+    protected static DeflaterFactory defaultDeflaterFactory = new DeflaterFactory();
+    protected static int defaultCompressionLevel = BlockCompressedStreamConstants.DEFAULT_COMPRESSION_LEVEL;
+
+    protected final BinaryCodec codec;
+    protected final byte[] uncompressedBuffer = new byte[BlockCompressedStreamConstants.DEFAULT_UNCOMPRESSED_BLOCK_SIZE];
+    protected int numUncompressedBytes = 0;
+    protected long mBlockAddress = 0;
+    protected File file = null;
+
+    // Really a local variable, but allocate once to reduce GC burden.
+    protected final byte[] singleByteArray = new byte[1];
+
+
+    /**
+     * Sets the default {@link DeflaterFactory} that will be used for all instances unless specified otherwise in the constructor.
+     * If this method is not called the default is a factory that will create the JDK {@link Deflater}.
+     * @param deflaterFactory non-null default factory.
+     */
+    public static void setDefaultDeflaterFactory(final DeflaterFactory deflaterFactory) {
+        if (deflaterFactory == null) {
+            throw new IllegalArgumentException("null deflaterFactory");
+        }
+        defaultDeflaterFactory = deflaterFactory;
+    }
+
+    public static DeflaterFactory getDefaultDeflaterFactory() {
+        return defaultDeflaterFactory;
+    }
+
+    /**
+     * Sets the GZip compression level for subsequent BlockCompressedOutputStream object creation
+     * that do not specify the compression level.
+     * @param compressionLevel 1 <= compressionLevel <= 9
+     */
+    public static void setDefaultCompressionLevel(final int compressionLevel) {
+        if (compressionLevel < Deflater.NO_COMPRESSION || compressionLevel > Deflater.BEST_COMPRESSION) {
+            throw new IllegalArgumentException("Invalid compression level: " + compressionLevel);
+        }
+        defaultCompressionLevel = compressionLevel;
+    }
+
+    public static int getDefaultCompressionLevel() {
+        return defaultCompressionLevel;
+    }
+
+    /**
+     * Prepare to compress at the given compression level
+     * @param file file to output
+     */
+    public AbstractBlockCompressedOutputStream(final File file) {
```

Comment:
Small code style thing. If you change the name of "file" parameter to "f" or something like that you can avoid usage of "this" in the constructor.